### PR TITLE
First draft of embedding management

### PR DESCRIPTION
**Pull request by jobergum**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+](https://medium.com/pinterest-engineering/searchsage-learning-search-query-representations-at-pinterest-654f2bb887fc) for
+different tasks without incurring the usual costs related to embedding
+versioning, storage, or inference infrastructure.
+
+
+![Vector Space Illustration](/assets/2023-03-31-tailoring-frozen-embeddings-with-vespa/image2.png)
+
+_A frozen model embeds data Q and D into vector space. By transforming
+the Q representation to Q’, the vector distance is reduced
+`(d(D,Q) < d(D,Q’)`. This figure illustrates fine-tuning of metric
```
Comment (geirst) [inline _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md:103]: Which distance is smallest? `d(D,Q)` or `d(D,Q’)`? Based on the description, it sounded like it should be `d(D,Q) > d(D,Q’)` instead.

File: _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+representations of items._
+
+With frozen embeddings, the data is embedded once using a foundational
+embedding model. Developers can then adapt the representation to
+specific tasks by adding transformation layers on the frozen
+embeddings. The following sections describe three different methods
+for adapting frozen embeddings for search or recommendation use
+cases and how these approaches are represented in Vespa.
+
+* Tuning the query tower in two-tower embedding models
```
Comment (geirst) [inline _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md:116]: There is something strange with these 3 sentences. Leftovers?

File: _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+again. By de-coupling the model weights of the query and document
+tower, developers can treat the document embeddings as frozen,
+to avoid re-encoding all the items. When fine-tuning the model for the
+specific task, developers can instead just tune the query tower and
+leave the frozen document tower alone.
+
+![Frozen Query Tower](/assets/2023-03-31-tailoring-frozen-embeddings-with-vespa/image3.png)
+
+The frozen document tower and embeddings significantly reduce the
+complexity and cost of serving and training. For training the query
```
Comment (geirst) [inline _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md:138]: Long and complex sentence starting with "For training the query...". Consider rewrite.

File: _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+tower, because there is no need to encode the document vector over
+again, the document’s embedding representation can be fetched
+directly from Vespa during training. This saves at least 2x of
+computational complexity during training as one does not need to
+run inference with the document tower during training. In practice,
```
Comment (geirst) [inline _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md:143]:
```suggestion
run inference with the document tower. In practice,
```

File: _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+since documents are longer than queries and Transformer models
+scales quadratic with input lengths, the computational saving is
+higher than 2x.
+
+On the serving side in Vespa, there is no need to re-process the
+frozen document embedding, this saves the compute of performing the
```
Comment (geirst) [inline _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md:149]:
```suggestion
frozen document embedding. This saves the compute of performing the
```

File: _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+tensor](https://docs.vespa.ai/en/tensor-user-guide.html#constant-tensors)
+file. Meanwhile, the `transform_query` function [performs a vector
+matrix
+product](https://docs.vespa.ai/playground/#N4KABGBEBmkFxgNrgmUrWQPYAd5QGNIAaFDSPBdDTAF30gDUBTA2rAJwFoBbAQ1ocAlgA8wODlgAmAVzYA6MADEhHAM60wAd2ZgpzaEIB2uvmABurdhzAADRrbB8jUp2H6DRdgOq24kMlQAX0Cg0gxqclwGZhJAiAp8SJpIIwZGOJpMWIRIWmYjNU4AChFEAGYAXQBKOEQABnkAVmIwRpa25sqArJCMMMDkhOjc2PCUyjR4zDTc70ysqByofMKSrUQAFkriMqraxEQARnl61pOj8-kAJh2ka9PWh8uwB9viY8ewC6v3z7PvvIXidbt1pn1guMIEMoCNCAtyJMYQl6LkAIIWKycXgCYRiCTSOSaIRqMC0AAWuh4MgANrQhDgaUICAIhFgjGAsNAyZTMWxOE4XO5cV5nK41DIeDxjABzTmWGwU3QELBS9l6IQ8ApqNlGfzg0JQqZZbCTSBjaaJKjTBKzKAIlLLSASnjFRhgABUYG8rS01R6NAhEAhlRAQSAA),
+returning a modified vector of the same dimensionality.
+
+This representation is used to score the documents in the first-phase
+ranking expression. Note that this is effectively represented as
+[a re-ranking phase](https://docs.vespa.ai/en/phased-ranking.html)
+as the query tensor used for the nearestNeighbor search is untouched.
```
Comment (geirst) [inline _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md:283]:
```suggestion
as the query tensor used for the [nearestNeighbor](https://docs.vespa.ai/en/reference/query-language-reference.html#nearestneighbor) search is untouched.
```

File: _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+to unlock true personalization.
+
+
+## Advanced transformation using Deep Neural Networks
+
+Another approach for customization is to use the query and document
+embeddings as input to another Deep Neural Network (DNN) model.
+This approach can be combined with the previously mentioned
+approaches because it's applied as a re-scoring model in a [phased
+ranking ](https://docs.vespa.ai/en/phased-ranking.html)pipeline.
```
Comment (geirst) [inline _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md:298]:
```suggestion
ranking](https://docs.vespa.ai/en/phased-ranking.html)pipeline.
```

File: _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+                  args=args,
+                  f="custom_similarity.onnx",
+                  input_names = input_names,
+                  output_names = output_names,
+                  opset_version=15)
+
+</pre>
+The above [PyTorch](https://pytorch.org/) [model.py](https://github.com/vespa-engine/sample-apps/blob/master/custom-embeddings/model.py) snippet defines a custom
+DNN-based similarity model which takes the query and document
+embedding as input. This model is exported to[ ONNX](https://onnx.ai/)
```
Comment (geirst) [inline _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md:335]:
```suggestion
embedding as input. This model is exported to [ONNX](https://onnx.ai/)
```

File: _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+format for [accelerated
+inference](https://blog.vespa.ai/stateful-model-serving-how-we-accelerate-inference-using-onnx-runtime/)
+using Vespa’s support for [ranking with
+ONNX](https://docs.vespa.ai/en/onnx.html) models.
+
+Notice that [PyTorch ](https://pytorch.org/)model uses batch
```
Comment (geirst) [inline _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md:341]:
```suggestion
Notice that the [PyTorch](https://pytorch.org/) model uses batch
```

**Review by geirst — APPROVED** [review]
👍 Very interesting!

File: _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md (hunk 1 above)
Comment (jobergum) [reply _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md:103]: Thanks!

File: _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+other schema fields. Adding the new vector field still adds to the serving cost,
+as we double the resource usage footprint related to indexing and storage. 
+
+* After all this, we are finally ready to evaluate the new embedding
+model online. Depending on the outcome of the online evaluation, we can
+garbage collect either the new or old embedding data. 
+
+That's a lot of complexity and cost to evaluate a model online, but now we can relax? 
+Wait, our PM now wants to introduce [news article recommendations](https://docs.vespa.ai/en/tutorials/news-4-embeddings.html)
+for the home page, and the ML team is planning on using embeddins for this project. We also hear they are
```
Comment (geirst) [inline _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md:80]:
```suggestion
for the home page, and the ML team is planning on using embeddings for this project. We also hear they are 
```

File: _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md (hunk 2 above)
Comment (geirst) [reply _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md:119]: Sorry, I didn't see this was a list. Ignore the comment.

File: _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+  } 
+  field embedding type tensor&lt;float&gt;(x[384]) { 
+    indexing: input text | embed frozen | attribute | index
+  }
+}
+</pre>
+
+In this case, Vespa will produce embeddings using a `frozen` embedding
+model, and at query time, we can either use the frozen model to
+encode the query or a new fine-tuned model. Deplying multipe query tower models allows
```
Comment (geirst) [inline _posts/2023-03-31-tailoring-frozen-embeddings-with-vespa.md:178]:
```suggestion
encode the query or a new fine-tuned model. Deploying multiple query tower models allows
```

**Review by geirst — COMMENTED** [review]
After reviewing the latest changes.

**Review by kkraune — APPROVED** [review]
💯

**Pull request merged**